### PR TITLE
WT-7817 make tiered storage address cookies backward compatible

### PR DESCRIPTION
**Pull request by keithbostic**
Move the file ID to the end of the address cookie and to the end of the checkpoint cookie.

This allows old releases to detect address/checkpoint cookies they cannot understand and fail gracefully, that is, they won't interpret the file ID as the block offset and drop core or return a corruption failure. Use the length of the address/checkpoint cookie to detect if there is a file ID included in the cookie (this means passing the cookie sizes around; we've always had those sizes in hand, we've just never needed them before). In all cases, assert the entire cookie is consumed.

Rename the block address cookie buffer-to-addr and addr-to-buffer functions as unpack and pack functions to match other parts of WiredTiger.

Don't do misplaced block checks, or verbose messages, when freeing blocks that were read from other objects. We can't free them, but we shouldn't do anything else, either.

Define WT_TIERED_CURRENT_ID as the maxium 4B value instead of a bit pattern.

Remove minor unnecessary error handling and scratch buffer logic from tiered storage functions.

### CODE REVIEW

**Comment by keithbostic** [comment]
@ddanderson, @fedorova, @sueloverso, @keitharnoldsmith: nobody replied  "No!" to my email, so I took that as encouragement and took a run at integrating tiered storage addresses with the historic address cookie layout.

If nobody likes this approach, we can simply delete it.

Sasha, I added you to the review list since you've probably looked at address cookies as much as anyone now. :)

File: src/block/block_addr.c [hunk 1]
```diff
@@ -37,12 +49,11 @@ __block_buffer_to_addr(WT_BLOCK *block, const uint8_t **pp, uint32_t *objectidp,
      * that's simpler than testing sizes of 0 all over the place.
      */
     if (s == 0) {
+        *objectidp = 0;
         *offsetp = 0;
-        *objectidp = *sizep = *checksump = 0;
+        *sizep = *checksump = 0;
```
Comment (sueloverso) [inline src/block/block_addr.c:54]: These are all `uint32_t *` types. Why pull the initialization of just `objectidp` out?

File: src/block/block_addr.c [hunk 2]
```diff
@@ -51,46 +62,72 @@ __block_buffer_to_addr(WT_BLOCK *block, const uint8_t **pp, uint32_t *objectidp,
...
-    uint64_t l, o, s, c;
+    uint64_t i, o, s, c;
 
-    /* See the comment above: this is the reverse operation. */
+    /* See the comment above about storing large offsets: this is the reverse operation. */
     if (size == 0) {
+        i = 0;
         o = WT_BLOCK_INVALID_OFFSET;
-        l = s = c = 0;
+        s = c = 0;
```
Comment (sueloverso) [inline src/block/block_addr.c:78]: Same here.

**Review by sueloverso — APPROVED** [review]
LGTM with a minor question. But @ddanderson is the one who understands this best. I will go with his choice of whether we want this or not. Seems reasonable to me.

**Comment by sueloverso** [comment]
However, the PR testing failed on compiling and clang-format type jobs so there is probably some minor cleanup needed somewhere.

File: src/block/block_addr.c (hunk 1 above)
Comment (keithbostic) [reply src/block/block_addr.c:54]: Habit, really -- setting them in the order they appear in the argument list, I find that way I don't worry about missing one.

**Review by ddanderson — APPROVED** [review]
I like this change, some nice improvements and code cleanup.

File: src/block/block_addr.c [hunk 3]
```diff
@@ -149,96 +184,126 @@ __wt_block_addr_string(
...
-      block, pp, &ci->root_objectid, &ci->root_offset, &ci->root_size, &ci->root_checksum));
-    WT_RET(__block_buffer_to_addr(
-      block, pp, &ci->alloc.objectid, &ci->alloc.offset, &ci->alloc.size, &ci->alloc.checksum));
-    WT_RET(__block_buffer_to_addr(
-      block, pp, &ci->avail.objectid, &ci->avail.offset, &ci->avail.size, &ci->avail.checksum));
-    WT_RET(__block_buffer_to_addr(block, pp, &ci->discard.objectid, &ci->discard.offset,
+    /*
+     * See the comment above about address cookies and sizes for an explanation.
+     *
+     * Note the passing of a address cookie size of 0, triggering the called function's branch to
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/block/block_addr.c]: Rewritten in 3102bbf3f.

File: src/block/block_addr.c [hunk 4]
```diff
@@ -9,23 +9,35 @@
...
 
-    if (block->has_objects)
-        WT_RET(__wt_vunpack_uint(pp, 0, &l));
-    else
-        l = 0;
+    /*
+     * Address cookies are a file offset, size and checksum triplet, with optional object ID: unpack
+     * the trailing object ID if there are bytes following the triple. The checkpoint cookie is more
+     * complicated: it has four address blocks all from the same file. Rather than retrofit the file
+     * ID into each of those address blocks (which would mean somehow figuring out the length of
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/block/block_addr.c]: Rewritten in 3102bbf3f.

File: src/block/block_addr.c [hunk 5]
```diff
@@ -149,96 +184,126 @@ __wt_block_addr_string(
...
 
-    WT_RET(__wt_block_addr_to_buffer(
-      block, pp, objectid, ci->root_offset, ci->root_size, ci->root_checksum));
-    WT_RET(__wt_block_addr_to_buffer(
-      block, pp, objectid, ci->alloc.offset, ci->alloc.size, ci->alloc.checksum));
+    /*
+     * See the comment above about address cookies and sizes for an explanation.
+     *
+     * Note passing a file object ID of 0, triggering the called function's branch to ignore file
+     * object IDs.
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/block/block_addr.c]: Rewritten in 3102bbf3f.

File: src/block/block_addr.c [hunk 6]
```diff
@@ -149,96 +184,126 @@ __wt_block_addr_string(
...
+    WT_RET(__block_addr_unpack(block, &ckpt, 0, &ci->discard.objectid, &ci->discard.offset,
       &ci->discard.size, &ci->discard.checksum));
-    WT_RET(__wt_vunpack_uint(pp, 0, &a));
+    WT_RET(__wt_vunpack_uint(&ckpt, 0, &a));
     ci->file_size = (wt_off_t)a;
-    WT_RET(__wt_vunpack_uint(pp, 0, &a));
+    WT_RET(__wt_vunpack_uint(&ckpt, 0, &a));
     ci->ckpt_size = a;
+    ci->root_objectid = ci->alloc.objectid = ci->avail.objectid = ci->discard.objectid = 0;
+    if (WT_PTRDIFF(*&ckpt, start) != ckpt_size) {
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/block/block_addr.c]: Fixed in 3102bbf3f.

File: src/block/block_addr.c [hunk 7]
```diff
@@ -51,46 +62,72 @@ __block_buffer_to_addr(WT_BLOCK *block, const uint8_t **pp, uint32_t *objectidp,
...
+    start = p;
+    WT_RET(__block_addr_unpack(block, &p, addr_size, objectidp, offsetp, sizep, checksump));
+
+    /*
+     * Assert the entire cookie was consumed.
+     *
+     * TODO: The assert prevents future extensions of the cookie this code should ignore, but that
+     * may be unavoidable, as the object ID is not a required field, extending the cookie may not be
+     * simple. Consider adding a flag byte before the optional object ID, describing the object ID
+     * and any other fields.
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/block/block_addr.c:127]: I'm a little hesitant, but as you say, it gives us options going forward (and allows us a reasonable way to assert the address we're cracking is exactly the address we think we're cracking, I'd rather fail in the address unpack than with some random corruption message).

@sueloverso, @fedorova, @keitharnoldsmith: does anybody object to this change?

File: src/block/block_addr.c [hunk 8]
```diff
@@ -51,46 +62,72 @@ __block_buffer_to_addr(WT_BLOCK *block, const uint8_t **pp, uint32_t *objectidp,
...
     WT_RET(__wt_vpack_uint(pp, 0, s));
     WT_RET(__wt_vpack_uint(pp, 0, c));
+
+    /*
+     * Don't store object IDs of zero, the function that cracks the cookie defaults IDs to 0.
+     *
+     * TODO: Ideally, don't store a object ID if there's only a single object. We want to be able to
+     * convert any existing object to a stack, and that means starting with a single object with no
+     * object IDs, where all future objects in the stack know a missing object ID is a reference to
+     * the base object.
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/block/block_addr.c:95]: @ddanderson, I think this is a discussion worth having in JIRA, I moved it to: https://jira.mongodb.org/browse/WT-7817?focusedCommentId=3926952&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-3926952.

**Comment by keithbostic** [comment]
@ddanderson, c990bb3 is the other part of the change for your consideration. There are two things going on here: first, fail with a verbose error message if we can't find a file referenced by the checkpoint/address cookie (so, historic release seeing tiered-storage file, or a missing file caused by some other error), and second, a step toward making tiered storage address cookies first-class citizens. It seems to me that anywhere we have `block->has_objects` we're better served by testing the address itself against the information we have about the block, but I'm not confident of that yet, I'd understand if you strongly disagreed.

File: src/block/block_addr.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (keitharnoldsmith) [reply src/block/block_addr.c:127]: No objections here.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM!

**Comment by keithbostic** [comment]
@ddanderson, I've pushed the changes to add the flag byte we discussed.

Also, did you ever look at this change?

> @ddanderson, [c990bb3](https://github.com/wiredtiger/wiredtiger/commit/c990bb3badac377a26f75ef47d46e416673d824c) is the other part of the change for your consideration. There are two things going on here: first, fail with a verbose error message if we can't find a file referenced by the checkpoint/address cookie (so, historic release seeing tiered-storage file, or a missing file caused by some other error), and second, a step toward making tiered storage address cookies first-class citizens. It seems to me that anywhere we have `block->has_objects` we're better served by testing the address itself against the information we have about the block, but I'm not confident of that yet, I'd understand if you strongly disagreed.

Also, do you have any thoughts on what testing we should do for release compatibility before we merge?

File: src/block/block_addr.c [hunk 9]
```diff
@@ -36,8 +39,23 @@ __block_addr_unpack(WT_BLOCK *block, const uint8_t **pp, size_t addr_size, uint3
...
+
+    /*
+     * If there's an address cookie size and either there is no flag value or the flag value is the
+     * object ID flag by itself, assert the cookie was entirely consumed. Future extensions will use
+     * different cookie flag values (although the file-ID flag might still be set, our test is for
+     * equality).
+     */
+    WT_ASSERT(session,
+      addr_size == 0 || (flags != 0 && flags != WT_BLOCK_COOKIE_FILEID) ||
+        WT_PTRDIFF(*pp, begin) == addr_size);
```
Comment (ddanderson) [inline src/block/block_addr.c:58]: Can we strengthen this assert?  Like:
```
    WT_ASSERT(session,
      addr_size == 0 || ((flags != 0 && flags != WT_BLOCK_COOKIE_FILEID) &&
        WT_PTRDIFF(*pp, begin) == addr_size));
```

File: src/block/block_addr.c [hunk 10]
```diff
@@ -207,34 +214,38 @@ __block_ckpt_unpack(WT_SESSION_IMPL *session, WT_BLOCK *block, const uint8_t *ck
...
-     * may be unavoidable, as the object ID is not a required field, extending the cookie may not be
-     * simple. Consider adding a flag byte before the optional object ID, describing the object ID
-     * and any other fields.
+     * If there is no flag value or the flag value is the object ID flag by itself, assert the
+     * cookie was entirely consumed. Future extensions will use different cookie flag values
+     * (although the file-ID flag might still be set, our test is for equality).
      */
-    WT_ASSERT(session, WT_PTRDIFF(ckpt, start) == ckpt_size);
+    WT_ASSERT(session,
+      (flags != 0 && flags != WT_BLOCK_COOKIE_FILEID) || WT_PTRDIFF(ckpt, begin) == ckpt_size);
```
Comment (ddanderson) [inline src/block/block_addr.c:248]: This doesn't seem right to me.  Isn't `WT_PTRDIFF(ckpt, start) == ckpt_size)` always true?

**Review by ddanderson — APPROVED** [review]
Some questions on the asserts.
Everything else looks good.

**Comment by ddanderson** [comment]
Yes, that looks good.  Thanks!

> Also, did you ever look at this change?
> 
> > @ddanderson, [c990bb3](https://github.com/wiredtiger/wiredtiger/commit/c990bb3badac377a26f75ef47d46e416673d824c) is the other part of the change for your consideration.

**Review by fedorova — APPROVED** [review]
LGTM

File: src/block/block_addr.c (hunk 9 above)
Comment (keithbostic) [reply src/block/block_addr.c:58]: @ddanderson, this doesn't look right to me.

The current assert reads "if we got an address size, then if there weren't any flags or the flags were flags this release knows about, then we should have read all of the address cookie".

I'd read your assert as "if we got an address size, then there must be flags, and they may not be 0 or the flags we know about, and we must have read all the address cookie"?

I think there's an extra set of parenthesis in there, too, so maybe that's not what you intended to type?

Or maybe I'm just misreading it!!

File: src/block/block_addr.c (hunk 10 above)
Comment (keithbostic) [reply src/block/block_addr.c:248]: If the flags value isn't one we know about, then it's a future release and we may not have consumed the fields added by that release.

Let me know if I'm missing something.

**Pull request merged**